### PR TITLE
Option to raise on failed responses

### LINKED ISSUE #55: Complain louder when num generic requests != num generic responses after parsing
And surface the errors to the user

### PR DESCRIPTION
**Pull request by RyanMarten**
Sometimes you are ok with some responses failing and just getting back the dataset. 

Other times, you are not ok with this. E.g. getting the responses to exactly all of the prompts in a dataset with a bunch of different models to compare. 

In that second case, we add an option that will raise an error if not all requests are completed. 

The best solution is to try to reduce the number requests that failed. To that end, I increased the default number of retries from 5 to 10. However, we should continue to track down bugs and figure out ways to mitigate failures. 

Fixes #55

### CODE REVIEW

**Comment by RyanMarten** [comment]
We are seeing lots of failed responses due to rate limit errors. 

root cause
- for deep infra, the rate limiting is determined by concurrent requests

potential solutions
- #252 
- #253

**Comment by RyanMarten** [comment]
Test (with 600 concurrent request limit)
```
python tests/simple_online.py --log-level DEBUG --model deepinfra/meta-llama/Llama-3.3-70B-Instruct  --n-requests 700 --max-requests-per-minute 100000000000 --max-tokens-per-minute 10000000000000 --max-retries 0
```

```
2024-12-13 13:07:25,236 - bespokelabs.curator.request_processor.base_online_request_processor - WARNING - Request 687 failed with Exception litellm.RateLimitError: RateLimitError: DeepinfraException - Error code: 429 - {'error': {'message': 'Model busy, retry later'}}, attempts left 0
2024-12-13 13:07:25,236 - bespokelabs.curator.request_processor.base_online_request_processor - ERROR - Request 687 failed permanently after exhausting all 0 retry attempts. Errors: ["litellm.RateLimitError: RateLimitError: DeepinfraException - Error code: 429 - {'error': {'message': 'Model busy, retry later'}}"]
2024-12-13 13:07:25,239 - bespokelabs.curator.request_processor.base_online_request_processor - WARNING - Request 675 failed with Exception litellm.RateLimitError: RateLimitError: DeepinfraException - Error code: 429 - {'error': {'message': 'Model busy, retry later'}}, attempts left 0
2024-12-13 13:07:25,239 - bespokelabs.curator.request_processor.base_online_request_processor - ERROR - Request 675 failed permanently after exhausting all 0 retry attempts. Errors: ["litellm.RateLimitError: RateLimitError: DeepinfraException - Error code: 429 - {'error': {'message': 'Model busy, retry later'}}"]
```


```
2024-12-13 13:07:44,726 - bespokelabs.curator.request_processor.base_online_request_processor - INFO - Status tracker: Tasks - Started: 700, In Progress: 0, Succeeded: 366, Failed: 334, Already Completed: 0
Errors - API: 0, Rate Limit: 0, Other: 334, Total: 334
2024-12-13 13:07:44,726 - bespokelabs.curator.request_processor.base_online_request_processor - WARNING - 334 / 700 requests failed. Errors logged to /Users/ryan/.cache/curator/b79d429f8149519e/responses_0.jsonl.
2024-12-13 13:07:44,744 - bespokelabs.curator.request_processor.base_request_processor - INFO - Finalizing writer
2024-12-13 13:07:44,746 - bespokelabs.curator.request_processor.base_request_processor - INFO - Read 700 responses.
2024-12-13 13:07:44,746 - bespokelabs.curator.request_processor.base_request_processor - WARNING - 334 requests failed
```

Related to #252 these errors are not correctly tracked as rate limit errors, they are tracked as "other". 

The new logged warning is as expected and returns partial dataset.

Now with `--require-all-responses`

```
python tests/simple_online.py --log-level DEBUG --model deepinfra/meta-llama/Llama-3.3-70B-Instruct  --n-requests 700 --max-requests-per-minute 100000000000 --max-tokens-per-minute 10000000000000 --max-retries 0 --require-all-responses
```

```
2024-12-13 13:13:21,822 - bespokelabs.curator.request_processor.base_online_request_processor - WARNING - Request 688 failed with Exception litellm.RateLimitError: RateLimitError: DeepinfraException - Error code: 429 - {'error': {'message': 'Model busy, retry later'}}, attempts left 0
2024-12-13 13:13:21,822 - bespokelabs.curator.request_processor.base_online_request_processor - ERROR - Request 688 failed permanently after exhausting all 0 retry attempts. Errors: ["litellm.RateLimitError: RateLimitError: DeepinfraException - Error code: 429 - {'error': {'message': 'Model busy, retry later'}}"]
```

Raises error as desired
```
2024-12-13 13:13:50,051 - bespokelabs.curator.request_processor.base_online_request_processor - INFO - Processing complete. Results saved to /Users/ryan/.cache/curator/b79d429f8149519e/responses_0.jsonl
2024-12-13 13:13:50,051 - bespokelabs.curator.request_processor.base_online_request_processor - INFO - Status tracker: Tasks - Started: 700, In Progress: 0, Succeeded: 360, Failed: 340, Already Completed: 0
Errors - API: 0, Rate Limit: 0, Other: 340, Total: 340
2024-12-13 13:13:50,051 - bespokelabs.curator.request_processor.base_online_request_processor - WARNING - 340 / 700 requests failed. Errors logged to /Users/ryan/.cache/curator/b79d429f8149519e/responses_0.jsonl.
2024-12-13 13:13:50,071 - bespokelabs.curator.request_processor.base_request_processor - INFO - Finalizing writer
2024-12-13 13:13:50,078 - bespokelabs.curator.request_processor.base_request_processor - INFO - Read 700 responses.
2024-12-13 13:13:50,078 - bespokelabs.curator.request_processor.base_request_processor - WARNING - 340 requests failed.
Traceback (most recent call last):
  File "/Users/ryan/curator/tests/simple_online.py", line 54, in <module>
    main(args)
  File "/Users/ryan/curator/tests/simple_online.py", line 25, in main
    dataset = prompter(dataset, batch_cancel=args.cancel)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/src/bespokelabs/curator/llm/llm.py", line 257, in __call__
    return self._completions(self._request_processor, dataset, working_dir, batch_cancel)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/src/bespokelabs/curator/llm/llm.py", line 356, in _completions
    dataset = request_processor.run(
              ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/src/bespokelabs/curator/request_processor/base_online_request_processor.py", line 233, in run
    return self.create_dataset_files(working_dir, parse_func_hash, prompt_formatter)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/src/bespokelabs/curator/request_processor/base_request_processor.py", line 335, in create_dataset_files
    raise ValueError(f"Some requests failed and require_all_responses is True")
ValueError: Some requests failed and require_all_responses is True
```

File: tests/simple_online.py [hunk 1]
```diff
@@ -41,5 +43,12 @@ def main(args):
     parser.add_argument(
         "--max-tokens-per-minute", type=int, help="Max tokens per minute", default=None
     )
+    parser.add_argument("--max-retries", type=int, help="Max retries", default=None)
+    parser.add_argument(
+        "--require-all-responses",
+        action="store_true",
```
Comment (madiator) [inline tests/simple_online.py:49]: Sorry what is this `store_true`? Can you not use bool?

File: src/bespokelabs/curator/utils.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# https://stackoverflow.com/questions/845058/how-to-get-the-line-count-of-a-large-file-cheaply-in-python
+# https://stackoverflow.com/a/68385697
+def _file_gen(reader):
+    b = reader(1024 * 1024)
+    while b:
+        yield b
+        b = reader(1024 * 1024)
+
+
+# instead of requiring counting lines, we can store metadata file that has the number of requests in each file
```
Comment (madiator) [inline src/bespokelabs/curator/utils.py:10]: Capitalize "instead"

File: src/bespokelabs/curator/utils.py
Comment (madiator) [inline src/bespokelabs/curator/utils.py:1]: In general, utils.py is not a good idea because it ends up becoming a magnet for all sorts of things. Can you move these to the right files or rename this file from utils.py to something more focused.

File: src/bespokelabs/curator/request_processor/openai_online_request_processor.py [hunk 3]
```diff
@@ -81,6 +81,8 @@ def __init__(
         frequency_penalty: Optional[float] = None,
         max_requests_per_minute: Optional[int] = None,
         max_tokens_per_minute: Optional[int] = None,
+        require_all_responses: bool = False,
```
Comment (madiator) [inline src/bespokelabs/curator/request_processor/openai_online_request_processor.py:84]: I think this should be True by default!

File: src/bespokelabs/curator/request_processor/openai_online_request_processor.py [hunk 4]
```diff
@@ -81,6 +81,8 @@ def __init__(
...
+        max_retries: Optional[int] = None,
```
Comment (madiator) [inline src/bespokelabs/curator/request_processor/openai_online_request_processor.py:85]: Can you document this? What happens if I set this to 0?

File: src/bespokelabs/curator/request_processor/base_online_request_processor.py [hunk 5]
```diff
@@ -22,8 +22,9 @@
 logger = logging.getLogger(__name__)
 logger.setLevel(logging.INFO)
 
-DEFAULT_REQUESTS_PER_MINUTE = 100
-DEFAULT_TOKENS_PER_MINUTE = 100_000
+DEFAULT_MAX_REQUESTS_PER_MINUTE = 100
```
Comment (madiator) [inline src/bespokelabs/curator/request_processor/base_online_request_processor.py:25]: Why is this renamed?
Does it mean that we may do less than max_reqeusts?

File: src/bespokelabs/curator/request_processor/base_online_request_processor.py [hunk 6]
```diff
@@ -124,8 +125,10 @@ def __init__(
         frequency_penalty: Optional[float] = None,
         max_requests_per_minute: Optional[int] = None,
         max_tokens_per_minute: Optional[int] = None,
+        require_all_responses: bool = False,
```
Comment (madiator) [inline src/bespokelabs/curator/request_processor/base_online_request_processor.py:128]: Default should be True, otherwise users are in for a surprise!

File: src/bespokelabs/curator/request_processor/base_online_request_processor.py [hunk 7]
```diff
@@ -134,13 +137,22 @@ def __init__(
...
-        manual_value = getattr(self, name)
-        default_value = getattr(self, f"DEFAULT_{name.upper()}")
+        if name == "max_requests_per_minute":
+            manual_value = self.max_requests_per_minute
+            default_value = DEFAULT_MAX_REQUESTS_PER_MINUTE
+        elif name == "max_tokens_per_minute":
+            manual_value = self.max_tokens_per_minute
+            default_value = DEFAULT_MAX_TOKENS_PER_MINUTE
+        else:
+            raise ValueError(f"Invalid rate limit name: {name}")
```
Comment (madiator) [inline src/bespokelabs/curator/request_processor/base_online_request_processor.py:154]: I think this is the right way to do this: you use properties or separate methods.

```
__init__(...
   max_requests_per_minute: Optional[int] = None
  ...
):
  self._max_requests_per_minute = max_requests_per_minute

@property
def max_requests_per_minute(self):
  if self._max_requests_per_minute:
    return self._max_requests_per_minute
  else:
     return DEFAULT_MAX_REQUESTS_PER_MINUTE
```

And then perhaps you can get rid of this `get_rate_limit` method and just access it: `self.max_requests_per_minute`.

Or if you need to process baed on header, you do this:

```
def max_requests_per_minute(self, header):
  header_rpm = ..
  ...
```

File: src/bespokelabs/curator/request_processor/base_online_request_processor.py [hunk 8]
```diff
@@ -472,7 +484,7 @@ async def handle_single_request_with_retries(
                 retry_queue.put_nowait(request)
             else:
                 logger.error(
-                    f"Request {request.task_id} failed permanently after exhausting all 5 retry attempts. "
+                    f"Request {request.task_id} failed permanently after exhausting all {self.max_retries} retry attempts. "
```
Comment (madiator) [inline src/bespokelabs/curator/request_processor/base_online_request_processor.py:487]: if require_all_responses is True, shouldn't we terminate here?

File: src/bespokelabs/curator/request_processor/base_request_processor.py [hunk 9]
```diff
@@ -319,14 +320,35 @@ def create_dataset_files(
...
+            n_requests = 0
+            for request_file in request_files:
+                n_requests += count_lines(request_file)
+
+            if n_requests != total_responses_count:
+                logger.warning(
+                    f"{n_requests - total_responses_count} requests do not have responses. n_requests is {n_requests} and n_responses is {total_responses_count}"
+                )
+                if self.require_all_responses:
+                    os.remove(dataset_file)
```
Comment (madiator) [inline src/bespokelabs/curator/request_processor/base_request_processor.py:348]: Why are you removing the file?

**Review by madiator — COMMENTED** [review]
Added some comments!

File: tests/simple_online.py (hunk 1 above)
Comment (RyanMarten) [reply tests/simple_online.py:49]: This stores a boolean

File: src/bespokelabs/curator/request_processor/openai_online_request_processor.py (hunk 4 above)
Comment (RyanMarten) [reply src/bespokelabs/curator/request_processor/openai_online_request_processor.py:85]: with 0 it tries only once

File: src/bespokelabs/curator/request_processor/base_online_request_processor.py (hunk 5 above)
Comment (RyanMarten) [reply src/bespokelabs/curator/request_processor/base_online_request_processor.py:25]: Just to be clearer. Everywhere else we are calling the rate limit max_requests_per_minute.

File: src/bespokelabs/curator/request_processor/base_online_request_processor.py (hunk 6 above)
Comment (RyanMarten) [reply src/bespokelabs/curator/request_processor/base_online_request_processor.py:128]: Changed it to None everywhere except for base_request_processor where it is True

File: src/bespokelabs/curator/request_processor/base_online_request_processor.py (hunk 8 above)
Comment (RyanMarten) [reply src/bespokelabs/curator/request_processor/base_online_request_processor.py:487]: This is a good question, I think it makes more sense to continue processing as many requests successfully as possible and then failing at the end.

E.g. we are running a data generation job on the cluster and we don't want to to proceeded to the next step (thus raise an error), but we do want it to fufill as many requests as possible so when we restart the job (and it resumes from where it left off) there is less to do.

File: src/bespokelabs/curator/request_processor/base_request_processor.py (hunk 9 above)
Comment (RyanMarten) [reply src/bespokelabs/curator/request_processor/base_request_processor.py:348]: the dataset file is removed because when it exists it means that the dataset was successfully created and we load from that (cache) instead of sending requests. 

Note: we aren't removing the responses file so when the program is run again curator will try to resume where it left off.

File: src/bespokelabs/curator/request_processor/base_online_request_processor.py (hunk 8 above)
Comment (madiator) [reply src/bespokelabs/curator/request_processor/base_online_request_processor.py:487]: Yeah I am ok with not terminating. I suppose we are mostly trying to give some feedback to the user.

**Review by madiator — APPROVED** [review]
